src/libraries/System.Private.CoreLib/src/System/Globalization/Normalization.Unix.cs

```diff
@@ -41,26 +47,60 @@ internal static string Normalize(string strInput, NormalizationForm normalizatio
 
             ValidateArguments(strInput, normalizationForm);
 
-            char[] buf = new char[strInput.Length];
-
-            for (int attempts = 2; attempts > 0; attempts--)
+            char[]? toReturn = null;
+            try
             {
-                int realLen = Interop.Globalization.NormalizeString(normalizationForm, strInput, strInput.Length, buf, buf.Length);
+                Span<char> buffer = strInput.Length <= 512
+                    ? stackalloc char[512]
+                    : (toReturn = ArrayPool<char>.Shared.Rent(strInput.Length));
 
-                if (realLen == -1)
+                for (int attempt = 0; attempt < 2; attempt++)
                 {
-                    throw new ArgumentException(SR.Argument_InvalidCharSequenceNoIndex, nameof(strInput));
+                    int realLen;
+                    fixed (char* pInput = strInput)
+                    fixed (char* pDest = &MemoryMarshal.GetReference(buffer))
+                    {
+                        realLen = Interop.Globalization.NormalizeString(normalizationForm, pInput, strInput.Length, pDest, buffer.Length);
+                    }
+
+                    if (realLen == -1)
+                    {
+                        throw new ArgumentException(SR.Argument_InvalidCharSequenceNoIndex, nameof(strInput));
+                    }
+
+                    if (realLen <= buffer.Length)
+                    {
+                        ReadOnlySpan<char> result = buffer.Slice(0, realLen);
+                        return result.SequenceEqual(strInput)
+                            ? strInput
+                            : new string(result);
+                    }
+
+                    Debug.Assert(realLen > 512);
```

Comment:
Nit: 512 is used several times throughput the method... put it into a const?

Reply:
Changed